tests/deploy/test_timer_handling.py

```diff
@@ -64,3 +144,155 @@ def assert_fired(timer: MockTimer, at: datetime.datetime) -> None:
     assert_fired(*t4)
     assert_fired(*t5)
     assert t3[0].activated_at is None
+
+
+@pytest.fixture
+def environment() -> uuid.UUID:
+    return uuid.UUID("83d604a0-691a-11ef-ae04-c8f750463317")
+
+
+@pytest.fixture
+def make_resource_minimal(environment):
+    def make_resource_minimal(
+        rid: ResourceIdStr,
+        values: dict[str, object],
+        requires: list[str],
+        status: state.ComplianceStatus = state.ComplianceStatus.HAS_UPDATE,
+    ) -> state.ResourceDetails:
+        """Produce a resource that is valid to the scheduler"""
+        attributes = dict(values)
+        attributes["requires"] = requires
+        character = json.dumps(
+            {k: v for k, v in attributes.items() if k not in ["requires", "provides", "version"]},
+            default=custom_json_encoder,
+            sort_keys=True,  # sort the keys for stable hashes when using dicts, see #5306
+        )
+        m = hashlib.md5()
+        m.update(rid.encode("utf-8"))
+        m.update(character.encode("utf-8"))
+        attribute_hash = m.hexdigest()
+
+        return state.ResourceDetails(resource_id=rid, attributes=attributes, attribute_hash=attribute_hash)
+
+    return make_resource_minimal
```

Comment:
Copied, I don't know where to put them in the common part

Suggestions welcome

Reply:
The whole concept of the method could use a refactor tbh. I can't immediately think of any low hanging fruit.